From 777336e47986d99c007847d7ae72f3db8279da5b Mon Sep 17 00:00:00 2001
From: Spottedleaf <Spottedleaf@users.noreply.github.com>
Date: Tue, 14 Aug 2018 21:42:10 -0700
Subject: [PATCH] Allow Blocks to be accessed via a long key

The key can be retrieved via methods Location#toBlockKey() and
Block#getBlockKey()

World provides lookup for blocks by long key via method World#getBlockAtKey(long)

The formatting for the key is as follows:

10 bit y|27 bit z|27 bit x

The y value is considered unsigned while z and x are considered two's complement

Y range: [0, 1023]
X, Z range: [-67 108 864, 67 108 863]

diff --git a/src/main/java/org/bukkit/Location.java b/src/main/java/org/bukkit/Location.java
index 4c2a269d..2719eecb 100644
--- a/src/main/java/org/bukkit/Location.java
+++ b/src/main/java/org/bukkit/Location.java
@@ -13,7 +13,6 @@ import org.jetbrains.annotations.Nullable;
 
 // Paper start
 import java.util.Collection;
-import java.util.Collections;
 import java.util.function.Predicate;
 import org.bukkit.entity.Entity;
 import org.bukkit.entity.LivingEntity;
@@ -578,6 +577,17 @@ public class Location implements Cloneable, ConfigurationSerializable {
         blockLoc.setZ(getBlockZ());
         return blockLoc;
     }
+
+    // Paper Start
+    /**
+     * @return The block key for this location's block location.
+     * @see Block#getBlockKey(int, int, int)
+     */
+    public long toBlockKey() {
+        return Block.getBlockKey(getBlockX(), getBlockY(), getBlockZ());
+    }
+    // Paper End
+
     /**
      * @return A new location where X/Y/Z are the center of the block
      */
diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index 1b0744ed..fa736b07 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -80,6 +80,38 @@ public interface World extends PluginMessageRecipient, Metadatable {
     @NotNull
     public Block getBlockAt(@NotNull Location location);
 
+    // Paper start
+    /**
+     * Gets the {@link Block} at the given block key
+     *
+     * @param key The block key. See {@link Block#getBlockKey()}
+     * @return Block at the key
+     * @see Block#getBlockKey(int, int, int)
+     */
+    @NotNull
+    public default Block getBlockAtKey(long key) {
+        int x = Block.getBlockKeyX(key);
+        int y = Block.getBlockKeyY(key);
+        int z = Block.getBlockKeyZ(key);
+        return getBlockAt(x, y, z);
+    }
+
+    /**
+     * Gets the {@link Location} at the given block key
+     *
+     * @param key The block key. See {@link Location#toBlockKey()}
+     * @return Location at the key
+     * @see Block#getBlockKey(int, int, int)
+     */
+    @NotNull
+    public default Location getLocationAtKey(long key) {
+        int x = Block.getBlockKeyX(key);
+        int y = Block.getBlockKeyY(key);
+        int z = Block.getBlockKeyZ(key);
+        return new Location(this, x, y, z);
+    }
+    // Paper end
+
     /**
      * Gets the y coordinate of the lowest block at this position such that the
      * block and all blocks above it are transparent for lighting purposes.
diff --git a/src/main/java/org/bukkit/block/Block.java b/src/main/java/org/bukkit/block/Block.java
index 708288e9..c20f903a 100644
--- a/src/main/java/org/bukkit/block/Block.java
+++ b/src/main/java/org/bukkit/block/Block.java
@@ -153,6 +153,76 @@ public interface Block extends Metadatable {
      */
     int getZ();
 
+    // Paper Start
+    /**
+     * Returns this block's coordinates packed into a long value.
+     * Computed via: {@code Block.getBlockKey(this.getX(), this.getY(), this.getZ())}
+     * @see Block#getBlockKey(int, int, int)
+     * @return This block's x, y, and z coordinates packed into a long value
+     */
+    public default long getBlockKey() {
+        return Block.getBlockKey(this.getX(), this.getY(), this.getZ());
+    }
+
+    /**
+     * Returns the specified block coordinates packed into a long value
+     * <p>
+     * The return value can be computed as follows:
+     * <br>
+     * {@code long value = ((long)x & 0x7FFFFFF) | (((long)z & 0x7FFFFFF) << 27) | ((long)y << 54);}
+     * </br>
+     * </p>
+     *
+     * <p>
+     * And may be unpacked as follows:
+     * <br>
+     * {@code int x = (int) ((packed << 37) >> 37);}
+     * </br>
+     * <br>
+     * {@code int y = (int) (packed >>> 54);}
+     * </br>
+     * <br>
+     * {@code int z = (int) ((packed << 10) >> 37);}
+     * </br>
+     * </p>
+     *
+     * @return This block's x, y, and z coordinates packed into a long value
+     */
+    public static long getBlockKey(int x, int y, int z) {
+        return ((long)x & 0x7FFFFFF) | (((long)z & 0x7FFFFFF) << 27) | ((long)y << 54);
+    }
+
+    /**
+     * Returns the x component from the packed value.
+     * @param packed The packed value, as computed by {@link Block#getBlockKey(int, int, int)}
+     * @see Block#getBlockKey(int, int, int)
+     * @return The x component from the packed value.
+     */
+    public static int getBlockKeyX(long packed) {
+        return (int) ((packed << 37) >> 37);
+    }
+
+    /**
+     * Returns the y component from the packed value.
+     * @param packed The packed value, as computed by {@link Block#getBlockKey(int, int, int)}
+     * @see Block#getBlockKey(int, int, int)
+     * @return The y component from the packed value.
+     */
+    public static int getBlockKeyY(long packed) {
+        return (int) (packed >>> 54);
+    }
+
+    /**
+     * Returns the z component from the packed value.
+     * @param packed The packed value, as computed by {@link Block#getBlockKey(int, int, int)}
+     * @see Block#getBlockKey(int, int, int)
+     * @return The z component from the packed value.
+     */
+    public static int getBlockKeyZ(long packed) {
+        return (int) ((packed << 10) >> 37);
+    }
+    // Paper End
+
     /**
      * Gets the Location of the block
      *
-- 
2.21.0

